text/0000-variadic.md

```diff
@@ -0,0 +1,287 @@
+- Feature Name: variadic
+- Start Date: 2017-08-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Support defining C-compatible variadic functions in Rust, via new intrinsics.
+Rust currently supports declaring external variadic functions and calling them
+from unsafe code, but does not support writing such functions directly in Rust.
+Adding such support will allow Rust to replace a larger variety of C libraries,
+avoid requiring C stubs and error-prone reimplementation of platform-specific
+code, improve incremental translation of C codebases to Rust, and allow
+implementation of variadic callbacks.
+
+# Motivation
+[motivation]: #motivation
+
+Rust can currently call any possible C interface, and export *almost* any
+interface for C to call. Variadic functions represent one of the last remaining
+gaps in the latter. Currently, providing a variadic function callable from C
+requires writing a stub function in C, linking that function into the Rust
+program, and arranging for that stub to subsequently call into Rust.
+Furthermore, even with the arguments packaged into a `va_list` structure by C
+code, extracting arguments from that structure requires exceptionally
+error-prone, platform-specific code, for which the crates.io ecosystem provides
+only partial solutions for a few target architectures.
+
+This RFC does not propose an interface intended for native Rust code to pass
+variable numbers of arguments to a native Rust function, nor an interface that
+provides any kind of type safety. This proposal exists primarily to allow Rust
+to provide interfaces callable from C code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+C code allows declaring a function callable with a variable number of
+arguments, using an ellipsis (`...`) at the end of the argument list. For
+compatibility, unsafe Rust code may export a function compatible with this
+mechanism.
+
+Such a declaration looks like this:
+
+```rust
+pub unsafe extern "C" fn func(arg: T, arg2: T2, ...) {
+    // implementation
+}
+```
+
+The `...` at the end of the argument list declares the function as variadic.
+The function must use `extern "C"`, and must use `unsafe`. To expose
+such a function as a symbol for C code to call directly, the function may want
+to use `#[no_mangle]` as well; however, Rust code may also pass the function to
+C code expecting a function pointer to a variadic function.
+
+Unsafe Rust code can also define a variadic closure:
+
+```rust
+let closure = |arg, arg2, ...| {
+    // implementation
+};
```

Comment:
You can't cast a closure to `extern "C" fn`, so I don't think supporting variadic closure is necessary in this RFC.

```
fn main() {
    unsafe {
        let _x = (|| {}) as extern fn(); //~ ERROR E0605
    }
}
```

Reply:
Plus, I think variadic functions should always be unsafe functions, and closure currently can't be unsafe.

Reply:
I didn't realize this; I'd assumed that [RFC 1558](https://github.com/rust-lang/rfcs/pull/1558) made this possible, but digging through it, it doesn't allow coercing to `extern fn`, only to `fn`. I filed https://github.com/rust-lang/rust/issues/44291 to allow coercing to `extern fn`, but in the meantime, I'll drop this for now and move it to an open for the future. Would have been nice for more convenient callbacks.